feat(api): add bookingUrl field to event types API v2 response

dhairyashiil opened the pull request:
this pr adds a new 'bookingUrl' field to the api v2 endpoints: 
1. /event-types
2. /event-types/{eventTypeId}

Videos: 

1. _normal user_



https://github.com/user-attachments/assets/a33e22a8-f84a-4853-88a8-d46fe863ec22




2. _org owner_



https://github.com/user-attachments/assets/d88c529c-7ed3-4cee-9f81-f1dfba756338




3. _org member_




https://github.com/user-attachments/assets/ee532944-705d-4e11-b661-4bd86ae9bc9a

dhairyashiil committed:
feat(api): add bookingUrl field to event types API v2 response

Add a new bookingUrl field to EventTypeOutput_2024_06_14 that contains
the full, correct booking URL for each event type. This fixes the issue
where companion app and Chrome extension were hardcoding cal.com URLs
instead of using organization subdomains.

Changes:
- Add bookingUrl field to EventTypeOutput_2024_06_14 type definition
- Update repository to include organization data when fetching users
- Add buildBookingUrl method to compute URL using getOrgFullOrigin
- Include bookingUrl in getResponseEventType return object

The booking URL is computed using the first user's organization slug
(if any) to generate the correct subdomain URL (e.g., i.cal.com/keith/30min
for organization users instead of cal.com/keith/30min).

dhairyashiil committed:
fix: include organization data in getEventTypeByIdWithHosts

Update getEventTypeByIdWithHosts to use usersInclude to fetch
organization slug data, ensuring bookingUrl is computed correctly
for the GET /v2/event-types/{eventTypeId} endpoint.

dhairyashiil committed:
test: add unit tests and E2E assertions for bookingUrl field

dhairyashiil committed:
fix(api-v2): correct bookingUrl format for org users and fix double slashes

Fix issues with bookingUrl field in event types API v2 response:
- Remove trailing slashes from base URLs to prevent double slashes
- Use profile.username instead of user.username for organization users
  (profile contains clean username without org suffix)
- Include user profiles in repository queries to access profile data
- Create local org-domains utility to replace @calcom/features dependency
  which isn't available in API v2 runtime

Changes:
- Add apps/api/v2/src/lib/org-domains.ts with getOrgFullOrigin function
  adapted from @calcom/features/ee/organizations/lib/orgDomains
- Update event-types.repository.ts to include profiles in usersInclude
- Update buildBookingUrl() to prioritize profile data when available
- Update unit tests to cover organization user profile scenarios

Fixes:
- Double slashes in URLs (e.g., http://localhost:3000//user/slug)
- Incorrect username format for org users (e.g., owner1-acme instead
  of owner1)
- Missing organization subdomain in booking URLs (e.g., should be
  http://acme.localhost:3000/owner1/30min not
  http://localhost:3000/owner1-acme/30min)

dhairyashiil committed:
deslop ai code

dhairyashiil committed:
handle empty username in both code and write test for it, use select instead of include

keithwillcode commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/output-event-types.service.ts:
```diff
@@ -374,6 +387,25 @@ export class OutputEventTypesService_2024_06_14 {
     });
   }
 
+  buildBookingUrl(users: UserWithOrganization[], slug: string): string {
```
This code lives in our core. We should reuse it instead of introducing repeated logic that needs to be maintained in multiple places.

keithwillcode replied:
There’s still logic happening here that should live in core

keithwillcode commented on apps/api/v2/src/lib/org-domains.ts:
```diff
@@ -0,0 +1,23 @@
+import { getEnv } from "@/env";
```
We should not recreate this code for the reasons mentioned above. Let’s reuse the core code. I understand it may need to be refactored to work well but that’s ok.

dhairyashiil committed:
feat(api-v2): reuse core org domain logic for event type bookingUrl

Export getOrgFullOrigin and subdomainSuffix from @calcom/platform-libraries/organizations
to reuse existing core logic instead of duplicating it in API v2. This ensures consistency
across the codebase and reduces maintenance burden.

- Export getOrgFullOrigin and subdomainSuffix from platform-libraries/organizations
- Update output-event-types.service to import from @calcom/platform-libraries/organizations
- Remove duplicate org-domains.ts file from API v2
- Update test mocks to use platform-libraries import path
- buildBookingUrl method now uses core getOrgFullOrigin function

This addresses feedback to reuse core code rather than introducing duplicate logic.

dhairyashiil committed:
test(api-v2): exclude bookingUrl from output comparison in e2e test

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into devin/1766678740-add-booking-url-to-event-types

dhairyashiil committed:
refactor(api-v2): reuse core logic for event type bookingUrl

Replace duplicated username/org extraction logic in buildBookingUrl with
core functions. Use getBookerBaseUrlSync from @calcom/platform-libraries/organizations
for base URL generation, and add enrichUserWithProfile method that follows
the same pattern as core's UserRepository.enrichUsersWithTheirProfiles.

The enrichment logic is applied synchronously since profiles are already
fetched by the repository, avoiding the need to make the service async.
This ensures consistency with core patterns while maintaining the existing
synchronous API contract.

- Replace getOrgFullOrigin with getBookerBaseUrlSync
- Add enrichUserWithProfile method following core enrichment pattern
- Simplify buildBookingUrl to use enriched user data
- Update tests to mock getBookerBaseUrlSync and verify enrichment flow

dhairyashiil commented on packages/platform/types/event-types/event-types_2024_06_14/outputs/event-type.output.ts:
```diff
@@ -490,6 +490,14 @@ export class EventTypeOutput_2024_06_14 extends BaseEventTypeOutput_2024_06_14 {
+    example: "https://cal.com/john-doe/30min",
+    format: "uri",
+  })
+  bookingUrl!: string;
```
not needed right now

dhairyashiil committed:
trailing slash cubic comment

dhairyashiil committed:
Merge branch 'main' into devin/1766678740-add-booking-url-to-event-types

dhairyashiil committed:
resolve merge conflicts

supalarry reviewed (changes requested):
Great job as always. I have some feedback in regards code organization and re-using existing code in our codebase. Let me know what you think.

dhairyashiil committed:
address review comments

dhairyashiil committed:
address review comments 2

dhairyashiil committed:
better code

dhairyashiil committed:
Updated the test to give the user an organization

dhairyashiil committed:
resolve merge conflicts

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/output-event-types.service.ts:
```diff
@@ -405,6 +438,34 @@ export class OutputEventTypesService_2024_06_14 {
+    return `${normalizedBaseUrl}/${username}/${slug}`;
+  }
+
+  private getUserMainProfile(user: EventTypeUser): UserProfile | undefined {
```
Instead of copying and pasting `getUserMainProfile` from `apps/api/v2/src/modules/users/services/users.service.ts` we need to dependency inject `UsersService` into this service and then access `usersService.getUserMainProfile`. If you need any help with this let me know.

supalarry reviewed (changes requested):
Almost there:

dhairyashiil committed:
address review comments again

dhairyashiil committed:
Revert "address review comments again"

This reverts commit 622ea3fc8dcd7f5c3113614afcfc86beadecca0e.

dhairyashiil committed:
allow both the full UserWithProfile and the partial EventTypeUser (from select queries) to use the same getUserMainProfile() method, eliminating code duplication without changing the performance-optimized repository queries.

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into devin/1766678740-add-booking-url-to-event-types

supalarry reviewed (changes requested):
Review:

dhairyashiil committed:
reveiw points

dhairyashiil committed:
Merge branch 'main' into devin/1766678740-add-booking-url-to-event-types

supalarry reviewed (approved):
Well done! 🚀
